### PR TITLE
Qute - scan "templates" directory in all application archives

### PR DESCRIPTION
**Pull request by mkouba**

### CODE REVIEW

**Comment by gastaldi** [comment]
How would that behave when you have a JAR with templates targeting another template engine (Freemarker for example)? Would those be understood as Qute templates as well?

**Comment by mkouba** [comment]
> How would that behave when you have a JAR with templates targeting another template engine (Freemarker for example)? Would those be understood as Qute templates as well?

First of all, it must be an indexed _application archive_, i.e. a file with a marker (`beans.xml`, jandex index, etc.). Furthermore, you can exclude any file from scanning (e.g. `*.ftl`) from any `templates` dir via `quarkus.qute.template-path-exclude`.

I think that it's reasonable to scan all the templates by default. Hm, do you expect many apps that will include the `quarkus-qute` extension and application archives with non-qute templates?

File: test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java [hunk 1]
```diff
@@ -164,12 +168,54 @@ public QuarkusUnitTest assertException(Consumer<Throwable> assertException) {
...
         return this;
     }
 
+    /**
+     * Customize the application root.
+     * 
+     * @param applicationRootConsumer
+     * @return self
+     */
+    public QuarkusUnitTest withApplicationRoot(Consumer<JavaArchive> applicationRootConsumer) {
```
Comment (mkouba) [inline test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java:188]: @stuartwdouglas These are the methods we talked about...

**Comment by mkouba** [comment]
The `RabbitMQConnectorIT` failure is unrelated:
```
Caused by: org.testcontainers.containers.ContainerFetchException: Failed to pull image: rabbitmq:3.9-management`
```

**Review by stuartwdouglas — APPROVED** [review]

**Comment by gastaldi** [comment]
> First of all, it must be an indexed _application archive_, i.e. a file with a marker (`beans.xml`, jandex index, etc.). Furthermore, you can exclude any file from scanning (e.g. `*.ftl`) from any `templates` dir via `quarkus.qute.template-path-exclude`.

Ah cool! Not urgent but maybe it would be nice to put this in the Qute guide just in case


> I think that it's reasonable to scan all the templates by default. Hm, do you expect many apps that will include the `quarkus-qute` extension and application archives with non-qute templates?

Realistically I don't think so, but who knows :)

**Review by gastaldi — APPROVED** [review]

**Comment by mkouba** [comment]
> > First of all, it must be an indexed _application archive_, i.e. a file with a marker (`beans.xml`, jandex index, etc.). Furthermore, you can exclude any file from scanning (e.g. `*.ftl`) from any `templates` dir via `quarkus.qute.template-path-exclude`.
> 
> Ah cool! Not urgent but maybe it would be nice to put this in the Qute guide just in case

+1 will add it later.

> 
> > I think that it's reasonable to scan all the templates by default. Hm, do you expect many apps that will include the `quarkus-qute` extension and application archives with non-qute templates?
> 
> Realistically I don't think so, but who knows :)

Yeah, you're right :D

**Comment by mklueh** [comment]
@mkouba first of all, many thanks, it works. However, I tried to override a generic template of my service module (where the mail sending logic is defined) with a concrete customized template of the actual application and I'm running into this issue:

```
Caused by: java.lang.IllegalStateException: Duplicate templates found:

	at io.quarkus.qute.deployment.QuteProcessor.analyzeTemplates(QuteProcessor.java:344)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:887)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

Is this default behavior intended? Seems overriding is currently impossible.

One could argue that you can define the templates only in the application module and pass the TemplateInstance to a worker class within the submodule, but in my case the class dealing with the templates is a standalone service with it's own REST Api. So for my two applications, sharing and reusability of the same submodule is pretty much diminished without ugly workarounds.

In my opinion, the following behavior would make sense:

1. Throw "Duplicate templates found:" when templates equal in sibling modules
2. (Don't throw "Duplicate templates found:" when templates equal in sibling modules AND parent - just use parent)
3. Take parent template if parent and child templates equal

Do you think it would be possible to change that?

**Comment by mkouba** [comment]
> Is this default behavior intended? Seems overriding is currently impossible.

@mklueh Yes, it's intentional and the reason why we decided not to support duplicate templates is that we want to keep things simple. And similar rules do the opposite. It probably makes sense for you but I wouldn't be surprised if other users would be confused, e.g. a template from the submodule can be opened in the IDE, but the changes are not visible.

If I understand your use case the submodule is used as a dependency and also as a standalone app?

**Comment by mklueh** [comment]
@mkouba understood the complexity part, however, I'm personally more confused as the behavior does not align with how application.properties (and probably other resources) behave, where you can override pretty much everything in your application.

To be more precise with my use case, I have an auth module for example, that handles user registration, password forgotten and user verification, all resulting in an email being sent out.

This module is used by two different applications that should have their own custom E-Mail templates, aligned with the look of the application, while reusing the auth behavior

**Comment by mkouba** [comment]
> I'm personally more confused as the behavior does not align with how application.properties (and probably other resources)...

I understand but I wouldn't compare templates to config which has overridable config sources by design and the ordering is well-defined and straightforward.

> This module is used by two different applications that should have their own custom E-Mail templates, aligned with the look of the application, while reusing the auth behavior

This seems to be a perfect fit for the [template inheritance](https://quarkus.io/guides/qute-reference#include_helper) where the shared module defines the "base" template and applications use this base and specify the overriden parts.

**Comment by mklueh** [comment]
@mkouba Thanks for your help. With templates, it works perfectly! :)

**Pull request merged**